src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -185,22 +214,36 @@ public void Add(string key, object value)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            EnsureWritable();
-            InnerDictionary.Add(key, value);
+            Upgrade();
+
+            var list = ((ListStorage)_storage);
+            list.Inner.Add(new KeyValuePair<string, object>(key, value));
         }
 
         /// <inheritdoc />
         public void Clear()
         {
-            EnsureWritable();
-            InnerDictionary.Clear();
+            Upgrade();
```

Comment:
Applies to all delete-ish methods below e.g. `Remove()` should return `false` in `EmptyStorage` case.